src/Ookii.Dialogs.Wpf/ProgressDialog.cs

```diff
@@ -804,6 +852,29 @@ private void _backgroundWorker_ProgressChanged(object sender, ProgressChangedEve
                     OnProgressChanged(new ProgressChangedEventArgs(e.ProgressPercentage, data.UserState));
                 }
             }
-        }    
+        }
+
+        /// <summary>
+        /// Used to retrieve services from the "Sender" event args.
+        /// </summary>
+        /// <param name="serviceType">
+        /// The service to retrieve. currently, only supports one of these:<br/>
+        /// <see cref="IProgress{T}"/> with <see cref="int"/>.<br/>
+        /// <see cref="IProgress{T}"/> with <see cref="string"/>.<br/>
+        /// <see cref="System.Threading.CancellationTokenSource"/>.<br/>
+        /// <see cref="System.Threading.CancellationToken"/>.<br/>
+        /// </param>
+        /// <returns>An object that can be casted to the requested service.</returns>
+        object IServiceProvider.GetService(Type serviceType)
```

Comment:
Interesting approach, but I think this is giving the `ProgressDialog` more responsibility that it should have. Exposing the `CancellationToken` via `ProgressChangedEventArgs` should be enough for passing it down to consumers.

Reply:
The alternative I can see is to overload of DoWorkEventArgs , maybe named DoCancellableWorkEventArgs, which will include the token.

Consider this async task method:

```c#
async Task SomeAsyncTask(CancellationToken token);
```

The way to register the event would look like this;

```c#
var dlg = new ProgressDialog();
dlg.DoWork += async (s,e) => await SomeAsyncTask( (s as DoCancellableWorkEventArgs).CancellationToken);
```

it's the other only way I can think to passing the cancellationToken to the work method as an argument.

Reply:
I see... The dialog exposes the original API of the `BackgroundWorker` directly so we can't extend it.

I'm thinking it might be worth breaking this dependency and create our own `DoWorkEventHandler`, `DoWorkEventArgs`, etc. hiding the `BackgroundWorker` completely.

Reply:
That's the problem: One way or another you need to cast something to something else to get the token, so:

- You either cast Sender to System.ComponentModel.IServiceProvider and do GetService<CancellationToken>();
- Or you cast DoWorkEventArgs to some Ookii.DoCancellableEventArgs and you take the token from some property.

Both options are sub-optimal, but the first one at least uses a type that is already available on System.

Reply:
Agreed on these two possible options. There's also a third option where we introduce our own delegate for the `DoWork` event which uses our own `EventArgs` which includes the `CancellationToken` and wouldn't need the cast. The downside is that [it would break everyone using the `ProgressDialog`](https://github.com/ookii-dialogs/ookii-dialogs-wpf/pull/57/files#diff-177d031e2de7645b3515f02065f883d61265d64dbabbfe391517d888178e381c) as they have to adjust their event registration and handlers to use the new delegate.

I put together a a couple of quick PRs just to see how it would look:

- [Add ProgressDialogDoWorkEventArgs extension (non-breaking) #56](https://github.com/ookii-dialogs/ookii-dialogs-wpf/pull/56) - second option
- [Add ProgressDialogDoWorkEventArgs extension (breaking) #57](https://github.com/ookii-dialogs/ookii-dialogs-wpf/pull/57) - third option

There's a trade-off in all of them, so we'll just have to pick one...

- The first option seems to be the most obscure as the user needs to cast the `sender` to `IServiceProvider` first then call `GetService` and ask for a `CancellationToken`
- The second option is simpler and less obscure than the first one - though still a little obscure, with a single cast, and the user is just unpacking additional event args and will find the `CancellationToken` where they would expect it to be
- The third option seems like the cleanest one as everything is clear, no casts needed, and the `CancellationToken` is accessible where you'd expect, however it breaks everyone using the `ProgressDialog` today, to introduce a `CancellationToken` that not everyone will leverage

I think the second option - although not optimal as you say - is the best one. Doesn't break anyone, can be shipped quickly, and enable the feature we want, with a small caveat that it requires one cast to access the `CancellationToken`.

What do you think?

Reply:
I do like the first option with IServiceProvider because it doesn't introduce new specific types, but since its quite apparent you don't like that option, I would go for option 2 then:  that is, PR #56 because it's not a breaking change.

Also, not sure If I would place ProgressDialogDoWorkEventArgs in Ookii.Core  instead of OoKii.Wpf

Reply:
Cheers. Yes, it comes down to preference for sure. I see `GetService` more of a DI container-kind-of-method, and `CancellationToken` is not something I'd normally ask a DI container for... That said, if it's useful to have a `GetService` to get a hold of `IProgress<int>`  or `IProgress<string>` then by all means let's have it.

There's no Ookii.Core yet, so we'll keep it in the WPF assembly for now. In fact, I'll just merge #56 and you can rebase and make any changes you need in this PR.